### ISSUE TITLE
Searching for arabic words

### ISSUE THREAD

**Issue by Ashamandi** [codemirror/CodeMirror#4784]
Hi,
Searching function does not find all matched words in case of Arabic words, it fails to match some letters like " ا  , أ , آ" and "ى ,  ي " which are considered equal and have to be matched

![image](https://user-images.githubusercontent.com/17547411/26867525-00c380e4-4b66-11e7-8095-e77e698724dc.png)

I have tested through Chrome v 58.0 and firefox developer edition v 54.0 (64 bit)

Thank you!

**Comment by marijnh** [comment]
Do I understand correctly that this is about unicode string normalization? I.e. two different strings denoting the same text, which should be [normalized](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/normalize) before comparing them?

**Comment by Ashamandi** [comment]
Yes, right that`s what I meant

**Comment by marijnh** [comment]
It appears that most browsers support [`String.normalize`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/normalize) nowadays, so it might be possible to make the search code normalize strings. The difficult thing there is associating offsets into the normalized string (which is what you'd get from searching after normalizing) with offsets in the original string (which you need to point at the match in the editor). There's already something like that happening for case-folding, but that can be done one character at a time, which isn't the case for unicode normalization.

If somebody wants to work on this, I'd be very grateful.

**Comment by marijnh** [comment]
(Maybe, though I'm not sure yet, normalizing to `NFD` form always decomposes code points, never combines them, so that might be safe to do one character at a time.)

**Comment by marijnh** [comment]
Attached patch introduces normalization using `String.prototype.normalize`, but while that does help with things like precomposed characters vs extending characters, it doesn't appear to have any effect on the cases you describe -- it doesn't normalize these characters, and it doesn't appear that JavaScript comes with a built-in way to do that.

One option would be to allow code to pass a custom normalization function, and to write language-specific normalizers (or maybe a good one already exists -- I wasn't able to find it). But it wouldn't be a drop-in fix -- you'd have to explicitly set it up. Would that help for you?

**Comment by Ashamandi** [comment]
Yes, you are right. we should add a custom normalization function for that I will try to implement it as npm package, is it will be ok?

**Comment by marijnh** [comment]
The search cursor does not currently support providing a custom normalize function, because of the issue of converting positions in the normalized string to positions in the original string. Can your normalization function work one character at a time, or would it need to see the character's context as well? Do you need normalization for regexp searches or only plain string searches?

**Comment by Ashamandi** [comment]
No, It will work for one char no need for its context and about your second question yes we need to apply normalization on both regexp and plain strings.

**Comment by marijnh** [comment]
Okay, that's not currently possible. If you want to work on making it possible, a patch would be welcome. The main problem to solve is that, if we find the (normalized) search pattern in the (normalized) editor content, we need to be able to find the corresponding position in the original content. To do that, you'd need to know something about what the normalizer did. One approach would be to have it return a data structure or function that represents the changes it makes, and can be used to do this position mapping.
